### PR TITLE
Speculative

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: server/text_generation_server/models/__init__.py [hunk 1]
```diff
@@ -97,6 +101,8 @@ def get_model(
     else:
         raise RuntimeError(f"Unknown dtype {dtype}")
 
+    SPECULATE = 2
```
Comment (OlivierDehaene) [inline server/text_generation_server/models/__init__.py:104]: Remove?

File: router/src/infer.rs [hunk 2]
```diff
@@ -523,18 +524,30 @@ fn send_responses(
...
-        text: generation.token_text,
-        logprob: generation.token_logprob,
-        special: generation.token_is_special,
+    let tokens: Vec<Token> = if let Some(tokens_) = generation.tokens {
+        tokens_
+            .ids
+            .into_iter()
+            .zip(tokens_.logprobs.into_iter())
+            .zip(tokens_.texts.into_iter())
+            .zip(tokens_.is_special.into_iter())
```
Comment (OlivierDehaene) [inline router/src/infer.rs:533]: `is_special` is empty for prompt tokens so this zip will not work.

File: server/text_generation_server/models/flash_causal_lm.py [hunk 3]
```diff
@@ -186,7 +192,9 @@ def from_pb(
 
             # Paged attention
             # Remove one as the first token des not have a past
-            total_tokens = input_length + max_new_tokens - 1
+            from text_generation_server.models import SPECULATE
+            speculative_length = SPECULATE
```
Comment (OlivierDehaene) [inline server/text_generation_server/models/flash_causal_lm.py:196]: Can we use a get here?

File: server/text_generation_server/models/flash_causal_lm.py [hunk 4]
```diff
@@ -473,13 +483,15 @@ def concatenate(cls, batches: List["FlashCausalLMBatch"]) -> "FlashCausalLMBatch
             total_batch_size += len(b)
             total_slots += len(b.slots)
             blocks += b.blocks
+            speculative_length = 0 if b.speculative_ids is None else b.speculative_ids.shape[1]
```
Comment (OlivierDehaene) [inline server/text_generation_server/models/flash_causal_lm.py:486]: I think speculative_ids can never be None here.

File: server/text_generation_server/models/flash_causal_lm.py [hunk 5]
```diff
@@ -752,28 +806,43 @@ def generate_token(
             del batch
             raise e
 
+        try:
+            out, speculative_logits = out.logits, out.speculative_logits
+        except Exception:
+            out = out
+            speculative_logits = None
```
Comment (OlivierDehaene) [inline server/text_generation_server/models/flash_causal_lm.py:813]: Maybe this can be refactored or at least we can add a comment/have a more precise except than the blanket `Exception`.

File: server/text_generation_server/models/flash_causal_lm.py [hunk 6]
```diff
@@ -752,28 +806,43 @@ def generate_token(
...
+                speculative_logits = (
+                    speculative_logits[batch.prefill_next_token_indices] if prefill_logprobs else speculative_logits
+                )
         else:
             next_token_logits = out
 
-        next_input_ids, next_token_logprobs, logprobs = batch.next_token_chooser(
-            batch.all_input_ids_tensor[:, : batch.max_seqlen], next_token_logits
+
+        from text_generation_server.models import SPECULATE
```
Comment (OlivierDehaene) [inline server/text_generation_server/models/flash_causal_lm.py:828]: Use a get like we do for the cache manager instead.

File: server/text_generation_server/models/flash_causal_lm.py [hunk 7]
```diff
@@ -864,13 +939,15 @@ def generate_token(
             batch.next_token_chooser.do_sample,
             batch.next_token_chooser.seeds,
             batch.top_n_tokens,
-            next_token_ids,
-            next_token_logprobs,
+            # next_token_ids,
+            # next_token_logprobs,
```
Comment (OlivierDehaene) [inline server/text_generation_server/models/flash_causal_lm.py:943]:
```suggestion
```

File: server/text_generation_server/models/flash_causal_lm.py [hunk 8]
```diff
@@ -881,41 +958,59 @@ def generate_token(
             do_sample,
             seed,
             top_n_tokens,
-            next_token_id,
-            next_token_logprob,
+            # next_token_id,
+            # next_token_logprob,
```
Comment (OlivierDehaene) [inline server/text_generation_server/models/flash_causal_lm.py:962]:
```suggestion
```

File: server/text_generation_server/models/flash_causal_lm.py [hunk 9]
```diff
@@ -881,41 +958,59 @@ def generate_token(
...
             # Evaluate stopping criteria
-            stop, reason = stopping_criteria(
-                next_token_id,
-                next_token_text,
-            )
 
-            if not stop:
-                stopped = False
+            left = 0
+            for j, next_token_id in enumerate(_next_token_ids):
```
Comment (OlivierDehaene) [inline server/text_generation_server/models/flash_causal_lm.py:986]: Maybe this could happen in the same for loop as above?

File: server/text_generation_server/models/flash_causal_lm.py [hunk 10]
```diff
@@ -881,41 +958,59 @@ def generate_token(
...
+                stop, reason = stopping_criteria(
+                    next_token_id,
+                    next_token_text,
```
Comment (OlivierDehaene) [inline server/text_generation_server/models/flash_causal_lm.py:989]: This is not correct.

File: proto/generate.proto [hunk 11]
```diff
@@ -135,43 +135,27 @@ message GeneratedText {
     optional uint64 seed = 4;
 }
 
-message PrefillTokens {
```
[the remark this answers is not in the document]
Comment (Narsil) [reply proto/generate.proto:138]: Done.

File: router/client/src/lib.rs [hunk 12]
```diff
@@ -10,7 +10,7 @@ pub use pb::generate::v1::HealthResponse;
 pub use pb::generate::v1::InfoResponse as ShardInfo;
 pub use pb::generate::v1::{
```
[the remark this answers is not in the document]
Comment (Narsil) [reply router/client/src/lib.rs:11]: Done.

**Comment by OlivierDehaene** [comment]
We are also missing:
- new prometheus metrics for the number of accepted speculative ids
- speculative tokens count in `next_batch`
- maybe refactor infer to have the same InferResponse as before.

File: router/src/infer.rs [hunk 13]
```diff
@@ -515,6 +515,7 @@ fn send_responses(
 
     let mut stopped = false;
 
+    tracing::info!("Generation: {:?}", generation);
```
[the remark this answers is not in the document]
Comment (Narsil) [reply router/src/infer.rs]: Done.

File: router/src/infer.rs (hunk 2 above)
Comment (Narsil) [reply router/src/infer.rs]: Prompt is handled in `Prefill` which doesn´t iter over them, so we're good.

File: server/text_generation_server/models/__init__.py (hunk 1 above)
Comment (Narsil) [reply server/text_generation_server/models/__init__.py]: Changed to proper handling.

File: server/text_generation_server/models/flash_causal_lm.py (hunk 10 above)
Comment (Narsil) [reply server/text_generation_server/models/flash_causal_lm.py:981]: Fixed.

**Comment by Narsil** [comment]
> We are also missing:
> 
>     * new prometheus metrics for the number of accepted speculative ids
> 
>     * speculative tokens count in `next_batch`
> 
>     * maybe refactor infer to have the same InferResponse as before.

Did all 3.

File: integration-tests/models/test_flash_llama.py [hunk 14]
```diff
@@ -43,7 +43,7 @@ async def test_flash_llama_all_params(flash_llama, response_snapshot):
         seed=0,
     )
 
-    assert response.details.generated_tokens == 5
+    assert response.details.generated_tokens == 10
```
Comment (OlivierDehaene) [inline integration-tests/models/test_flash_llama.py:46]: It seems that the test is failing again

File: server/text_generation_server/cli.py [hunk 15]
```diff
@@ -137,6 +138,29 @@ def download_weights(
...
+            model_id = config["base_model_name_or_path"]
+            revision = "main"
+            try:
+                utils.weight_files(model_id, revision, extension)
+                logger.info(f"Files for parent {model_id} are already present on the host. " "Skipping download.")
+                return
+            # Local files not found
+            except (utils.LocalEntryNotFoundError, FileNotFoundError, utils.EntryNotFoundError):
+                pass
+        except (utils.LocalEntryNotFoundError, utils.EntryNotFoundError):
```
Comment (OlivierDehaene) [inline server/text_generation_server/cli.py:161]: Should this be the same as line 159?

File: server/text_generation_server/utils/tokens.py [hunk 16]
```diff
@@ -146,6 +147,21 @@ def from_pb(
...
+    dtype = input_ids.dtype
+    # speculative_ids = torch.zeros((B, speculate), device=device, dtype=dtype)
+    seeds = next_ids[accepted_ids.cumsum(dim=-1) -1 ]
+    indices = (input_ids == seeds.unsqueeze(-1)).max(dim=1).indices + 1
+    all_indices = indices.unsqueeze(-1).expand(B, speculate) + torch.arange(speculate, device=device)
+    all_indices = torch.clamp(all_indices, max=input_ids.shape[1] - 1)
+
+    # logger.info(f"All indices {all_indices} - {input_ids.shape}")
+    speculative_ids = input_ids.gather(dim=-1, index=all_indices)
+    return speculative_ids
```
Comment (OlivierDehaene) [inline server/text_generation_server/utils/tokens.py:164]:
```suggestion
    B = accepted_ids.shape[0]
    device = input_ids.device
    dtype = input_ids.dtype
    seeds = next_ids[accepted_ids.cumsum(dim=-1) -1 ]
    indices = (input_ids == seeds.unsqueeze(-1)).max(dim=1).indices + 1
    all_indices = indices.unsqueeze(-1).expand(B, speculate) + torch.arange(speculate, device=device)
    all_indices = torch.clamp(all_indices, max=input_ids.shape[1] - 1)

    speculative_ids = input_ids.gather(dim=-1, index=all_indices)
    return speculative_ids
```

File: server/text_generation_server/utils/tokens.py [hunk 17]
```diff
@@ -215,20 +231,93 @@ def __init__(
...
+            if speculative_scores is not None:
+                # Medusa provided some scores
+                speculative_ids = Greedy()(speculative_scores)
+            else:
+                # n-gram
+                speculative_ids = create_n_gram_speculation(input_ids, next_ids, accepted_ids, speculate, verbose)
+        else:
+            speculative_ids = None
+        # if verbose:
+        #     logger.info(f"new speculative ids {datetime.datetime.now() - start}")
```
Comment (OlivierDehaene) [inline server/text_generation_server/utils/tokens.py:318]:
```suggestion
```

File: server/text_generation_server/utils/tokens.py [hunk 18]
```diff
@@ -215,20 +231,93 @@ def __init__(
...
+            accepted_ids = torch.tensor(accepted_ids, device=input_ids.device, dtype=input_ids.dtype)
+            next_ids = next_ids[indices]
+            scores = scores[indices]
+            indices = torch.arange(B, device=input_ids.device) * S
+            if speculative_scores is not None:
+                speculative_scores = speculative_scores[indices + accepted_ids - 1]
+        else:
+            accepted_ids = torch.ones_like(next_ids)
+        # if verbose:
+        #     logger.info(f"Indices/accepted id {datetime.datetime.now() - start}")
```
Comment (OlivierDehaene) [inline server/text_generation_server/utils/tokens.py:303]:
```suggestion
```

File: server/text_generation_server/utils/tokens.py [hunk 19]
```diff
@@ -215,20 +231,93 @@ def __init__(
...
         self.device = device
 
-    def __call__(self, input_ids: torch.Tensor, scores: torch.Tensor):
-        if self.watermark_processor is not None:
-            scores = self.watermark_processor(input_ids, scores)
-        if self.repetition_processor is not None:
-            scores = self.repetition_processor(input_ids, scores)
+    def __call__(self, input_ids: torch.Tensor, scores: torch.Tensor, speculate: int, speculated_ids: Optional[torch.Tensor] = None, speculative_scores: Optional[torch.Tensor] = None, verbose=False):
+        import datetime
+        # from loguru import logger
```
Comment (OlivierDehaene) [inline server/text_generation_server/utils/tokens.py:236]:
```suggestion
```

File: server/text_generation_server/utils/tokens.py [hunk 20]
```diff
@@ -215,20 +231,93 @@ def __init__(
...
+                _scores = warper(input_ids, _scores)
+
+
+            _next_ids = self.choice(_scores)
+            scores[:, j] = _scores
+            next_ids[:, j] = _next_ids
+        next_ids = next_ids.view(B*S)
+        scores = scores.view( B* S, -1)
+        # if verbose:
+        #     logger.info(f"Scores {datetime.datetime.now() - start}")
```
Comment (OlivierDehaene) [inline server/text_generation_server/utils/tokens.py:270]:
```suggestion
        all_next_ids = []
        next_ids = torch.zeros((B, S), device=scores.device, dtype=torch.long)
        for j in range(S):
            _scores = scores[:, j]
            if self.watermark_processor is not None:
                _scores = self.watermark_processor(input_ids, _scores)
            if self.repetition_processor is not None:
                _scores = self.repetition_processor(input_ids, _scores)

            for warper in self.warpers:
                _scores = warper(input_ids, _scores)


            _next_ids = self.choice(_scores)
            scores[:, j] = _scores
            next_ids[:, j] = _next_ids
        next_ids = next_ids.view(B*S)
        scores = scores.view( B* S, -1)
```

**Comment by Narsil** [comment]
Sorry you reviewed this, I was making adjustements to reduce the overhead. Cleaned it up.

**Comment by Narsil** [comment]
@OlivierDehaene Good for review this time, I'll run a few benches.

**Review by OlivierDehaene — DISMISSED** [review]

**Comment by shcho1118** [comment]
I know this PR has been merged, but I have a question.
Where can I find the part about generating masking for tree-based attention, which is one of Medusa's features?
image

**Comment by DoubleVII** [comment]
> I know this PR has been merged, but I have a question. Where can I find the part about generating masking for tree-based attention, which is one of Medusa's features? Or does it just use top-1 from each medusa head (in which case it would be the same as causal)? image

As you say, this PR just use top-1 from each medusa head.
